### PR TITLE
Add validating webhook for AppDeployment

### LINKED ISSUE #1681: [Feature] add AppDeployment conditioned status
when i use AppDeployment and my AppDeployment file contains  an undeclared application version, kubevela-vela-core will logs error like :
```
{"level":"error","ts":1621275528.0065122,"logger":"controller","msg":"Reconciler error","reconcilerGroup":

"core.oam.dev","reconcilerKind":"AppDeployment","controller":"appdeployment","name":"example-

appdeploy","namespace":"default","error":"ApplicationRevision.core.oam.dev \"example-app-v2\" not 

found","stacktrace":"...."
```

### would like
reference type AppStatus add status description field  like runtimev1alpha1.ConditionedStatus，and add  the results of reconcile .

### alternatives

when apply AppDeployment，validate appRevisions .

### config yaml
```
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: example-app
  annotations:
    app.oam.dev/revision-only: "true"
spec:
  components:
    - name: testsvc
      type: webservice
      properties:
        addRevisionLabel: true
        image: crccheck/hello-world
        port: 8000
---
apiVersion: core.oam.dev/v1beta1
kind: AppDeployment
metadata:
  name: example-appdeploy
spec:
  appRevisions:
    - revisionName: example-app-v10
      placement:
        - distribution:
            replicas: 2

    - revisionName: example-app-v2
      placement:
        - distribution:
            replicas: 2

```

### PR DESCRIPTION
**Pull request by cnbailian**
fix #1681 

Current validation logic:
1. spec.apprevisions.[].RevisionName cannot be empty.
2. spec.apprevisions.[].RevisionName locate target must exist.

anything else to add?

### CODE REVIEW

**Review by wonderflow — COMMENTED** [review]
good job, let's write an e2e test for the webhook?

File: pkg/webhook/core.oam.dev/v1alpha2/appdeployment/validating_handler.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+			appRevisionName := appRevision.RevisionName
+			if len(appRevisionName) == 0 {
+				allErrs = append(allErrs, field.Required(fldPath.Child(fmt.Sprintf("[%d]", i)),
+					"target application revision name cannot be empty"))
+				continue
+			}
+
+			targetAppRevision := &v1beta1.ApplicationRevision{}
+			if err := h.Get(context.Background(), ktypes.NamespacedName{Namespace: appDeployment.Namespace, Name: appRevisionName},
+				targetAppRevision); err != nil {
```
Comment (lxd5866) [inline pkg/webhook/core.oam.dev/v1alpha2/appdeployment/validating_handler.go:102]: assess the err is notfond error type

File: pkg/webhook/core.oam.dev/v1alpha2/appdeployment/validating_handler.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+				allErrs = append(allErrs, field.NotFound(fldPath.Child("revisionName"), appRevisionName))
+				continue
```
Comment (lxd5866) [inline pkg/webhook/core.oam.dev/v1alpha2/appdeployment/validating_handler.go:104]: unnecessary continue

File: pkg/webhook/core.oam.dev/v1alpha2/appdeployment/validating_handler.go (hunk 1 above)
Comment (cnbailian) [reply pkg/webhook/core.oam.dev/v1alpha2/appdeployment/validating_handler.go:102]: yes, I will fix it

File: pkg/webhook/core.oam.dev/v1alpha2/appdeployment/validating_handler.go (hunk 2 above)
Comment (cnbailian) [reply pkg/webhook/core.oam.dev/v1alpha2/appdeployment/validating_handler.go:104]: sure

**Comment by cnbailian** [comment]
@wonderflow @lxd5866 Please review.

**Comment by wonderflow** [comment]
Thanks for your contribution. I'm sorry that I'm going to close it as AppDeployment is deprecated in the v1.2 release. You're very welcome to contribute for other issues.

Thanks again!

**Pull request closed without merging**